File: diis.py
import numpy as np
import argparse

def diis_xtrap(x_list, diis_resid):

    vec_dim, diis_dim = np.shape(x_list)
    B_dim = diis_dim + 1
    B = -1.0*np.ones((B_dim,B_dim))

    for i in range(diis_dim):
        for j in range(diis_dim):
            B[i,j] = np.dot(diis_resid[:,i].T,diis_resid[:,j])
    B[-1,-1] = 0.0

    rhs = np.zeros(B_dim)
    rhs[-1] = -1.0

    coeff = np.linalg.solve(B,rhs)
    x_xtrap = np.zeros(vec_dim)
    for i in range(diis_dim):
        x_xtrap += coeff[i]*x_list[:,i]

    return x_xtrap

def main(args):

    n = args.matrix_size
    maxiter = args.maxit
    omega = args.omega
    diis_size = args.diis_size
    tol = args.tolerance
    sparsity = args.sparsity
    flag_diis = args.diis

    A = np.zeros((n,n))
    b = np.random.rand(n)

    print('Building matrix...')
    for i in range(n):
        for j in range(n):
            if i == j:
                A[i,i] = float(i+1) + sparsity*np.random.rand(1)
            else:
                A[i,j] = sparsity*np.random.rand(1)

    x = b
    Az = A*(np.ones((n,n)) - np.eye(n))
    x_xtrap = np.zeros(n)
    x_list = np.zeros((n,diis_size))
    diis_resid = np.zeros((n,diis_size))

    res = np.dot(A,x) - b

    print('Jacobi solver...')
    it = 0
    while np.linalg.norm(res) > tol and it < maxiter:

        u = b - np.dot(Az,x)
        for j in range(n):
            x[j] = (1.0-omega)*x[j] + omega*u[j]/A[j,j]

        res = np.dot(A,x) - b

        if flag_diis:
            diis_resid[:,np.mod(it,diis_size)] = res
            x_list[:,np.mod(it,diis_size)] = x
            if it > diis_size+1:
                x = diis_xtrap(x_list,diis_resid)

        print('Iter - {}    Residuum = {}'.format(it,np.linalg.norm(res)))

        it += 1

    #print('|x - x_exact| = {}'.format(np.linalg.norm(x-x_exact)))

    return



if __name__ == '__main__':
    
    parser = argparse.ArgumentParser()
    parser.add_argument('-n','--matrix_size',type=int,help='Size of matrix',default=2000)
    parser.add_argument('-m','--maxit',type=int,help='Maximum number of jacobi iterations',default=100)
    parser.add_argument('-w','--omega',type=np.float64,help='Weighted jacobi damping parameter',default=1.0)
    parser.add_argument('-d','--diis_size',type=int,help='DIIS subspace dimension',default=5)
    parser.add_argument('-t','--tolerance',type=np.float64,help='Convergence tolerance',default=1e-08)
    parser.add_argument('-s','--sparsity',type=np.float64,help='Matrix sparsity',default=1e-03)
    parser.add_argument('--diis',type=bool,help='Boolean indicating whether DIIS is used',default=False)
    args = parser.parse_args()
    main(args)

File: ccf90/diis.f90
module diis

      use dgemm_module, only: dot

      implicit none

      contains

                subroutine diis_xtrap(X_xtrap, X_list, diis_resid_list)

                        real, intent(in) :: X_list(:,:), diis_resid_list(:,:)
                        real, intent(inout) :: X_xtrap(size(X_list,1))
                        real :: B(size(X_list,2)+1,size(X_list,2)+1), rhs(size(X_list,2)+1), coeff(size(X_list,2)+1)
                        integer :: i, j, k, diis_size, vec_dim, B_dim

                        vec_dim = size(X_list,1)
                        diis_size = size(X_list,2)
                        B_dim = diis_size + 1

                        B = 0.0
                        B(B_dim,:) = -1.0
                        B(:,B_dim) = -1.0
                        B(B_dim,B_dim) = 0.0

                        do i = 1,diis_size
                           do j = i,diis_size
                              B(i,j) = dot(diis_resid_list(:,i),diis_resid_list(:,j))
                              B(j,i) = B(i,j)
                           end do
                        end do

                        rhs = 0.0
                        rhs(B_dim) = -1.0
                        call solvegauss(B,rhs,B_dim,coeff)
                        X_xtrap = 0.0
                        do k = 1,diis_size
                           X_xtrap = X_xtrap + coeff(k)*X_list(:,k)
                        end do
                end subroutine diis_xtrap                                        


                subroutine solvegauss(A,b,N,x)

                        integer, intent(in) :: N
                        real, intent(inout) :: A(N,N), b(N)
                        real, intent(out) :: x(N)
                        real :: m, xsum
                        integer :: i, j

                        do j = 1,N-1
                           do i = N,j+1,-1
                              m = A(i,j)/A(j,j)
                              A(i,:) = A(i,:) - m*A(j,:)
                              b(i) = b(i) - m*b(j)
                           end do
                        end do

                        x = 0.0
                        x(N) = b(N)/A(N,N)
                        do i = N-1,1,-1
                           xsum = 0.0
                           do j = N,i+1,-1
                              xsum = xsum + A(i,j)*x(j)
                           end do
                           x(i) = (b(i)-xsum)/(A(i,i))
                        end do

                end subroutine solvegauss



end module diis
